processing/src/test/java/io/druid/segment/IndexMakerTest.java

```diff
@@ -0,0 +1,272 @@
+    map2.put(DIMS.get(0), nullList);
+    map2.put(DIMS.get(1), "dim10");
+
+
+    final Map<String, Object> map3 = new HashMap<>();
+    map3.put(DIMS.get(0), ImmutableList.<String>of("dim00", "dim01"));
+
+    final Map<String, Object> map4 = new HashMap<>();
+
+    final Map<String, Object> map5 = new HashMap<>();
+    map5.put(DIMS.get(1), "dim10");
+
+    final Map<String, Object> map6 = new HashMap<>();
+    map6.put(DIMS.get(1), null);
+
+
+    return Iterables.<Object[]>concat(
+        permute(map1)
+        ,permute(map1, map4)
+        ,permute(map1, map5)
+        ,permute(map5, map6)
+        ,permute(map4, map5)
+        ,Iterables.transform(ImmutableList.of(Arrays.asList(map1, map2, map3, map4, map5, map6)), OBJECT_MAKER)
+    );
+
+  }
+
+  private final Collection<InputRow> events;
+
+  public IndexMakerTest(
+      final Collection<InputRow> events
+  )
+  {
+    this.events = events;
+  }
+
+  IncrementalIndex toPersist;
+  File tmpDir;
+  File persistTmpDir;
+
+  @Before
+  public void setUp() throws IOException
+  {
+    toPersist = new OnheapIncrementalIndex(
+        JodaUtils.MIN_INSTANT,
+        QueryGranularity.NONE,
+        DEFAULT_AGG_FACTORIES,
+        1000000
+    );
+    for (InputRow event : events) {
+      toPersist.add(event);
+    }
+    tmpDir = Files.createTempDir();
+    persistTmpDir = Files.createTempDir();
+    IndexMerger.persist(toPersist, persistTmpDir, INDEX_SPEC);
+  }
+
+  @After
+  public void tearDown() throws IOException
+  {
+    try {
+      FileUtils.deleteDirectory(tmpDir);
+    }
+    finally {
+      FileUtils.deleteDirectory(persistTmpDir);
+    }
+  }
+
+  @Test
+  public void testSimpleReprocess() throws IOException
+  {
+    final IndexableAdapter adapter = new QueryableIndexIndexableAdapter(IndexIO.loadIndex(persistTmpDir));
+    Assert.assertEquals(events.size(), adapter.getNumRows());
+    reprocessAndValidate(persistTmpDir, tmpDir);
+  }
+
+  private File reprocessAndValidate(File inDir, File tmpDir) throws IOException
+  {
+    final File outDir = IndexMaker.convert(
+        inDir,
+        tmpDir,
+        INDEX_SPEC
+    );
+    IndexIO.DefaultIndexIOHandler.validateTwoSegments(outDir, persistTmpDir);
+    return outDir;
+  }
+
+  private File appendAndValidate(File inDir, File tmpDir) throws IOException
+  {
+    final File outDir = IndexMerger.append(
+        ImmutableList.<IndexableAdapter>of(new QueryableIndexIndexableAdapter(IndexIO.loadIndex(inDir))),
+        tmpDir,
+        INDEX_SPEC
+    );
+    IndexIO.DefaultIndexIOHandler.validateTwoSegments(outDir, persistTmpDir);
+    return outDir;
+  }
+
+  @Test
+  public void testIdempotentReprocess() throws IOException
+  {
+    final IndexableAdapter adapter = new QueryableIndexIndexableAdapter(IndexIO.loadIndex(persistTmpDir));
+    Assert.assertEquals(events.size(), adapter.getNumRows());
+    reprocessAndValidate(persistTmpDir, tmpDir);
+
+    final File tmpDir2 = Files.createTempDir();
```

Comment:
to make your cleanup easier and the code more readable, why not just reuse the temp directory created as part of setup, so you don't have to worry about deleting stuff in your test. We shouldn't fail the test because we can't delete temp files.